### ISSUE TITLE
[inductor] `F.gelu()` outputs accuracy inconsistency

### ISSUE THREAD

**Issue by shaoyuyoung** [pytorch/pytorch#142224]
### 🐛 Describe the bug

This occurs **both** on CPU and CUDA.

When using `cudagraphs` as backend, there is **NO** inconsistency.

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

torch.manual_seed(42)


class Model(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        x = F.gelu(x)
        return x


m = Model().cuda()
x = torch.randn(64, 64).cuda()

with torch.no_grad():
    m = m.eval()
    c_m = torch.compile(m)  # if backend="cudagraphs", there is no inconsistency

    y = m(x)
    c_y = c_m(x)

    print(torch.allclose(y, c_y))
    print(torch.max(torch.abs(y - c_y)))

```

### Error logs

on Inductor
```
False
tensor(2.3842e-07, device='cuda:0')
```

on cudagraphs
```
True
tensor(0., device='cuda:0')
```

### Versions

PyTorch version: 2.6.0.dev20241205+cu124
OS: Ubuntu 20.04.6 LTS (x86_64)

<details>
<summary>click for detailed env</summary>

```
PyTorch version: 2.6.0.dev20241205+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: 16.0.1
CMake version: version 3.26.0
Libc version: glibc-2.31

Python version: 3.12.7 | packaged by Anaconda, Inc. | (main, Oct  4 2024, 13:27:36) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-200-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.6.68
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB

Nvidia driver version: 560.35.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      40 bits physical, 48 bits virtual
CPU(s):                             20
On-line CPU(s) list:                0-19
Thread(s) per core:                 1
Core(s) per socket:                 20
Socket(s):                          1
NUMA node(s):                       1
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              85
Model name:                         Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz
Stepping:                           7
CPU MHz:                            2500.000
BogoMIPS:                           5000.00
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          640 KiB
L1i cache:                          640 KiB
L2 cache:                           80 MiB
L3 cache:                           16 MiB
NUMA node0 CPU(s):                  0-19
Vulnerability Gather data sampling: Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:        KVM: Vulnerable
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:             Mitigation; IBRS
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; IBRS; IBPB conditional; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI SW loop, KVM SW loop
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch topoext cpuid_fault invpcid_single pti ssbd ibrs ibpb fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat umip pku ospke avx512_vnni

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] optree==0.13.1
[pip3] pytorch-triton==3.2.0+git35c6c7c6
[pip3] torch==2.6.0.dev20241205+cu124
[pip3] torchaudio==2.5.0.dev20241205+cu124
[pip3] torchvision==0.20.0.dev20241205+cu124
[pip3] triton==3.0.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] nvidia-cublas-cu12        12.4.5.8                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.4.127                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.2.1.3                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.5.147               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.6.1.9                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.3.1.170               pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.6.2                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.4.127                 pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.4.127                 pypi_0    pypi
[conda] optree                    0.13.1                   pypi_0    pypi
[conda] pytorch-triton            3.2.0+git35c6c7c6          pypi_0    pypi
[conda] torch                     2.6.0.dev20241205+cu124          pypi_0    pypi
[conda] torchaudio                2.5.0.dev20241205+cu124          pypi_0    pypi
[conda] torchvision               0.20.0.dev20241205+cu124          pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi
```

</details>

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @aakhundov

**Comment by ezyang** [comment]
On main, on A100, I see no discrepancy:

```
$ cat > rrpo.py
import torch
import torch.nn as nn
import torch.nn.functional as F

torch.manual_seed(42)


class Model(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        x = F.gelu(x)
        return x


m = Model().cuda()
x = torch.randn(64, 64).cuda()

with torch.no_grad():
    m = m.eval()
    c_m = torch.compile(m)  # if backend="cudagraphs", there is no inconsistency

    y = m(x)
    c_y = c_m(x)

    print(torch.allclose(y, c_y))
    print(torch.max(torch.abs(y - c_y)))
(/home/ezyang/local/a/pytorch-env) [ezyang@devgpu005.nha1 ~/local/a/pytorch (a08e81a8)]$ python rrpo.py
True
tensor(0., device='cuda:0')
```

However, you are testing on a nightly. So potentially there is something specific to V100.

torch.compile reserves the right to change numerics slightly, as long as it leads to a result that is closer to the true value. Here is a modified script that compares against f64 reference

```
import torch
import torch.nn as nn
import torch.nn.functional as F

torch.manual_seed(42)


class Model(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        x = F.gelu(x)
        return x


m = Model().cuda()
x = torch.randn(64, 64).cuda()
x64 = x.to(torch.float64)

with torch.no_grad():
    m = m.eval()
    c_m = torch.compile(m)  # if backend="cudagraphs", there is no inconsistency

    y = m(x)
    c_y = c_m(x)

    y64 = m(x64)

    print(torch.max(torch.abs(y - y64)))
    print(torch.max(torch.abs(c_y - y64)))
```

I am curious to see what the result is on your hardware.

**Comment by shaoyuyoung** [comment]
Hi, @ezyang , thank you for your attention. After running your script, here is my log:
```
tensor(2.5206e-07, device='cuda:0', dtype=torch.float64)
tensor(2.5206e-07, device='cuda:0', dtype=torch.float64)
```

Do I need to try my original script on main again to find out if it's a version issue or a device issue?

Feel free to ask me to provide any assistance if it is helpful to this issue :)

**Comment by shaoyuyoung** [comment]
I have run my script on main (torch2.5.1), the log is below
```
False
tensor(2.3842e-07, device='cuda:0')
```

It's worth noting that when I set `atol=1.3e-6` and `rtol=1e-5`, `torch.allclose` returns **True**.

Because I see the code here.
https://github.com/pytorch/pytorch/blob/4a6c056466b990e1ceb6b45aa8d66236ca675c2d/torch/testing/_comparison.py#L68

Out of curiosity, should I use this tolerance in fp32? How did this tolerance come about? :)

**Comment by ezyang** [comment]
Black magic :P You might get some clues from the linked reference

**Comment by eellison** [comment]
IIRC, floating point gives 1 ULP error per floating point operation. 1 ULP == ~ 1.6777216e07

```
 ===== Forward graph 0 =====
 /data/users/eellison/pytorch/torch/fx/_lazy_graph_module.py class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "f32[20][1]cuda:0"):
        mul: "f32[20][1]cuda:0" = torch.ops.aten.mul.Tensor(arg0_1, 0.5)
        mul_1: "f32[20][1]cuda:0" = torch.ops.aten.mul.Tensor(arg0_1, 0.7071067811865476);  arg0_1 = None
        erf: "f32[20][1]cuda:0" = torch.ops.aten.erf.default(mul_1);  mul_1 = None
        add: "f32[20][1]cuda:0" = torch.ops.aten.add.Tensor(erf, 1);  erf = None
        mul_2: "f32[20][1]cuda:0" = torch.ops.aten.mul.Tensor(mul, add);  mul = add = None
        return (mul_2,)
```

There are 5 ops here. 1.6777216e-7 * 5 == 8.388608e-07. The error you have here is less than the margin for error. 

Yea, I would recommend using a less strict value for testing.
